apps/aecore/src/aec_sync.erl

```diff
@@ -230,132 +203,198 @@ terminate(_, _) ->
+get_next_work_item(ST) ->
+    lager:info("Nothing to do: ~p", [ST]),
+    {take_a_break, ST}.
+
+do_update_sync_task({done, PeerId}, ST = #sync_task{ chain = Chain }) ->
+    Chain1 = Chain#{ peers := maps:get(peers, Chain) -- [PeerId] },
+    ST#sync_task{ chain = Chain1 };
+do_update_sync_task({error, PeerId}, ST = #sync_task{ chain = Chain }) ->
+    Chain1 = Chain#{ peers := maps:get(peers, Chain) -- [PeerId] },
+    ST#sync_task{ chain = Chain1 }.
+
+get_sync_task(STId, #state{ sync_tasks = STs }) ->
+    case lists:keyfind(STId, #sync_task.id, STs) of
+        false -> {error, not_found};
+        ST    -> {ok, ST}
+    end.
+
+set_sync_task(ST = #sync_task{ id = STId }, State) ->
+    set_sync_task(STId, ST, State).
+
+set_sync_task(STId, ST, S = #state{ sync_tasks = STs }) ->
+    S#state{ sync_tasks = lists:keystore(STId, #sync_task.id, STs, ST) }.
+
+delete_sync_task(STId, S = #state{ sync_tasks = STs }) ->
+    S#state{ sync_tasks = lists:keydelete(STId, #sync_task.id, STs) }.
+
+add_chain_info(Chain = #{ chain_id := CId }, S) ->
+    {ok, ST = #sync_task{ chain = Chain2 }} = get_sync_task(CId, S),
+    ST1 = ST#sync_task{ chain = merge_chains(Chain, Chain2) },
+    set_sync_task(ST1, S).
+
+init_sync_task(Chain) ->
+    #sync_task{ id = maps:get(chain_id, Chain), chain = Chain }.
+
+merge_chains(#{ chain_id := CId, peers := Ps1, chain := C1 },
+             #{ chain_id := CId, peers := Ps2, chain := C2 }) ->
+    %% We sort descending...
+    Cmp = fun(#{ height := H1 }, #{ height := H2 }) -> H1 >= H2 end,
+    #{ chain_id => CId, peers => lists:usort(Ps1 ++ Ps2),
+       chain => lists:merge(Cmp, C1, C2) }.
+
+match_tasks(_Chain, [], []) ->
+    no_match;
+match_tasks(Chain, [], Acc) ->
+    {N, #{ chain_id := CId, peers := Peers }} = hd(lists:reverse(Acc)),
+    {inconclusive, Chain, {get_header, CId, Peers, N}};
+match_tasks(Chain1, [ST = #sync_task{ chain = Chain2 } | STs], Acc) ->
+    case match_chains(maps:get(chain, Chain1), maps:get(chain, Chain2)) of
+        equal     -> {match, ST};
+        different -> match_tasks(Chain1, STs, Acc);
+        {fst, N}  -> match_tasks(Chain1, STs, [{N, Chain1} | Acc]);
+        {snd, N}  -> match_tasks(Chain1, STs, [{N, Chain2} | Acc])
+    end.
+
+match_chains([#{ height := N1 } | C1], [#{ height := N2 } | _] = C2) when N1 > N2 ->
+    match_chains(C1, C2);
+match_chains([#{ height := N1 } | _] = C1, [#{ height := N2 } | C2]) when N1 < N2 ->
+    match_chains(C1, C2);
+match_chains([], [#{ height := N } | _]) ->
+    {snd, N};
+match_chains([#{ height := N } | _], []) ->
+    {fst, N};
+match_chains([#{ hash := H } | _], [#{ hash := H } | _]) ->
+    equal;
+match_chains(_, _) ->
+    different.
+
+do_handle_worker({new_worker, PeerId, Pid}, ST = #sync_task{ workers = Ws }) ->
+    case lists:keyfind(PeerId, 1, Ws) of
+        false    -> ok;
+        {_, Old} -> lager:info("Peer ~p already has a worker (~p)", [ppp(PeerId), Old])
+    end,
+    erlang:monitor(process, Pid),
```

Comment:
The worker should be linked rather than monitored so that the workers are brought down if the server goes down. That would also mean that the server needs to trap exits (but not the workers).

Since the workers would die if the server dies, there is probably no need for the `terminate/1` function to clean anything up.

An alternative is to add the workers to a simple_one_for_one supervisor, but the only benefit I can see is to avoid the full exit signal from the server ending up in the crash message for the workers.